src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+
+        Expr mutated = IRMutator::visit(op);
+        op = mutated.as<Call>();
+        if (!op) {
+            return mutated;
+        }
+
+        auto rewrite = IRMatcher::rewriter(op, op->type);
+        if (rewrite(intrin(Call::abs, widening_sub(x, y)), cast(op->type, intrin(Call::absd, x, y))) ||
+            false) {
+            return rewrite.result;
+        }
+
+        if (no_overflow(op->type)) {
+            // clang-format off
+            if (rewrite(halving_add(x + y, 1), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x, y + 1), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x + 1, y), rounding_halving_add(x, y)) ||
+                rewrite(halving_add(x - y, 1), rounding_halving_sub(x, y)) ||
+                rewrite(halving_sub(x + 1, y), rounding_halving_sub(x, y)) ||
+                rewrite(shift_right(x + y, 1), halving_add(x, y)) ||
+                rewrite(shift_right(x - y, 1), halving_sub(x, y)) ||
+                rewrite(rounding_shift_right(x + y, 1), rounding_halving_add(x, y)) ||
+                rewrite(rounding_shift_right(x - y, 1), rounding_halving_sub(x, y)) ||
+                false) {
+                return mutate(rewrite.result);
+            }
+            // clang-format on
+        }
+
+        if (op->is_intrinsic(Call::widening_mul)) {
+            internal_assert(op->args.size() == 2);
+            Expr narrow_a = strip_widening_cast(op->args[0]);
+            Expr narrow_b = strip_widening_cast(op->args[1]);
+            if (narrow_a.defined() && narrow_b.defined()) {
+                return mutate(Cast::make(op->type, widening_mul(narrow_a, narrow_b)));
+            }
+        } else if (op->is_intrinsic(Call::widening_add)) {
+            internal_assert(op->args.size() == 2);
+            for (halide_type_code_t t : {op->type.code(), halide_type_uint}) {
+                Type narrow_t = op->type.narrow().narrow().with_code(t);
+                Expr narrow_a = lossless_cast(narrow_t, op->args[0]);
+                Expr narrow_b = lossless_cast(narrow_t, op->args[1]);
+                if (narrow_a.defined() && narrow_b.defined()) {
+                    return mutate(Cast::make(op->type, widening_add(narrow_a, narrow_b)));
+                }
+            }
+        } else if (op->is_intrinsic(Call::widening_sub)) {
+            internal_assert(op->args.size() == 2);
+            for (halide_type_code_t t : {op->type.code(), halide_type_uint}) {
+                Type narrow_t = op->type.narrow().narrow().with_code(t);
+                Expr narrow_a = lossless_cast(narrow_t, op->args[0]);
+                Expr narrow_b = lossless_cast(narrow_t, op->args[1]);
+                if (narrow_a.defined() && narrow_b.defined()) {
+                    return mutate(Cast::make(op->type, widening_sub(narrow_a, narrow_b)));
+                }
+            }
+        }
+
+        if (op->is_intrinsic(Call::shift_right) || op->is_intrinsic(Call::shift_left)) {
```

Comment:
Why do we switch from else if to if here?

Reply:
It doesn't matter, just conceptually breaks things up by "category" of operation. I can change it if you want.